### ISSUE TITLE
Text-rendering texture mod

### ISSUE THREAD

**Issue by Warr1024** [minetest/minetest#9946]
## Problem

I want to be able to render text onto textures, e.g. to add a custom nametag to an entity that is properly occluded by terrain, only visible on one face, etc.  Using a bitmap font almost works, but it lacks proper unicode support, so I can't use it reliably for i18n player-supplied data or process translations.

## Solutions

A `[text:<x>,<y>:<size>:<escaped-string>` texture modifier that renders text using whatever font the user has configured.  The `<size>` parameter should adjust the font size to provide a reasonably consistent metric for size (e.g. line height in pixels) independent of the user's font choices, to provide a decent chance the string will fit into the texture.

Ideally, process the string for translations before rendering.

## Alternatives

Using a bitmap font with [combine might _almost_ work, except that i18n/unicode support would have to be done manually for each character, or e.g. names would have to be tightly constrained to the supported character set.

Other than that, if this can be done without engine changes, I'm open to suggestion, but I think having the engine provide support for this would offer the best result.

**Comment by rubenwardy** [comment]
Also see #1367

**Comment by MoNTE48** [comment]
Bounty 500 USD for it. 😉

**Comment by LoneWolfHT** [comment]
Is that a typo or potential mad rush to implement this?

**Comment by MoNTE48** [comment]
@LoneWolfHT sorry?

**Comment by LoneWolfHT** [comment]
You're going to pay $500 (Five hundred dollars) for a small feature like this?

**Comment by rubenwardy** [comment]
He's referring to #1367

**Comment by MoNTE48** [comment]
> You're going to pay $500 (Five hundred dollars) for a small feature like this?

Small bounty for a small feature. Are you ready to do it? 🙂

**Comment by LoneWolfHT** [comment]
Ah, looking at the site it seems like it's $20. That's more reasonable lol
Guess I'll peek, I want this feature too

**Comment by sfan5** [comment]
Note that the dicussion in #1367 is worth reading.
My preferred solution to node text display would be to have an overlay, rather than letting the client rendering text onto an existing image.

**Comment by rubenwardy** [comment]
This is much much easier to implement than the other issue, because it asks for a specific implementation

**Comment by MoNTE48** [comment]
I would clarify that I am interested in a "dynamic" text. Which will be stored in the node metadata.
As proposed in #1367. This will be useful for creating signs, blackboards, store ads ("Diamonds, price 3 coin"), etc.

**Comment by MoNTE48** [comment]
Instead of storing hundreds of textures to support Latin and Cyrillic, as well as plain text, italics, bold, small+large letters (~60*2*3=360+punctuation) I’m even afraid to mention how many characters in Japanese. We already have a predefined font that has all these characters. Saving a meta with 30 letters, specifying color and formatting costs almost nothing compared to the current form (I store texture modifiers in nodemeta and it takes up VERY much memory).

**Comment by LoneWolfHT** [comment]
@sfan5 what would be your recommended method for entities?

**Comment by sfan5** [comment]
To reuse the text node overlay (once it exists) for entities too.

**Comment by Warr1024** [comment]
It's pretty close.

Rendering to an image server-side is less efficient than just sending the text string would be.

Translation support would also be a mess.  It would mean generating a separate image for each client language in use, and sending different images to each client.  I don't know if MT will let you have multiple separate images but tell each client they're the same filename so that you can have a single entity appear with the correct language for each user...

**Comment by erlehmann** [comment]
>  I don't know if MT will let you have multiple separate images but tell each client they're the same filename so that you can have a single entity appear with the correct language for each user...

I believe @appgurueu might be able to answer if and how that is possible – we had a similar topic on IRC recently.

**Comment by rubenwardy** [comment]
>  I don't know if MT will let you have multiple separate images but tell each client they're the same filename so that you can have a single entity appear with the correct language for each user...

Yes, this is possible - you can send textures to a specific player using `to_player` in `dynamic_add_media`. This means that you can get the language from `get_player_information` and then translate and send the correct texture to each player individually

**Comment by Warr1024** [comment]
Nevermind, it turns out the provided mod also doesn't solve the problem because it's restrictively licensed.

I think it should be implied that issues opened for MT imply that the solution should be at least as free as MT itself.  Unless you are willing to relicense as LGPL or more free, no, it doesn't satisfy the original request.

As for the translation thing, it was originally a nice-to-have, so yes, a mod that implements the same features would probably be satisfactory, assuming it worked.

**Comment by erlehmann** [comment]
> Nevermind, it turns out the provided mod also doesn't solve the problem because it's restrictively licensed.

What exactly is the issue with the license here? I.e. which use case does the AGPLv3 hinder?

**Comment by rollerozxa** [comment]
> > Nevermind, it turns out the provided mod also doesn't solve the problem because it's restrictively licensed.
> 
> What exactly is the issue with the license here? I.e. which use case does the AGPLv3 hinder?

It makes it impossible to use the mod library in a game without tainting it and turning the entire game AGPLv3. Whether that's a good or a bad thing is up for interpretation.

**Comment by rubenwardy** [comment]
Depending on your intrepretation of (A)GPL license, it's also impossible to use it from a mod that isn't AGPL

**Comment by erlehmann** [comment]
> Bounty 500 USD for it. 😉

@MoNTE48 is this offer still valid? :money_mouth_face:
